sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubAsyncClient.java

```diff
@@ -0,0 +1,1073 @@
+                tryEmitNext(disconnectedEventSink, new DisconnectedEvent(connectionId, null));
+
+                handleNoRecovery().subscribe(null, thr -> {
+                    logger.atWarning()
+                        .log("Failed to auto reconnect session: " + thr.getMessage());
+                });
+            } else {
+                handleRecovery().timeout(RECOVER_TIMEOUT, Mono.defer(() -> {
+                    // client should be RECOVERING, after timeout
+                    clientState.changeState(WebPubSubClientState.DISCONNECTED);
+                    // send DisconnectedEvent
+                    tryEmitNext(disconnectedEventSink, new DisconnectedEvent(connectionId, null));
+
+                    // fallback
+                    return handleNoRecovery();
+                })).subscribe(null, thr -> {
+                    logger.atWarning()
+                        .log("Failed to recover or reconnect session: " + thr.getMessage());
+                });
+            }
+        }
+    }
+
+    private void handleMessage(Object webPubSubMessage) {
+        if (logger.canLogAtLevel(LogLevel.VERBOSE)) {
+            try {
+                String json = JacksonAdapter.createDefaultSerializerAdapter()
+                    .serialize(webPubSubMessage, SerializerEncoding.JSON);
+                logger.atVerbose().addKeyValue("message", json).log("Received message");
+            } catch (IOException e) {
+                throw logger.logExceptionAsError(
+                    new UncheckedIOException("Failed to serialize received message for VERBOSE logging", e));
+            }
+        }
+
+        if (webPubSubMessage instanceof GroupDataMessage) {
+            GroupDataMessage groupDataMessage = (GroupDataMessage) webPubSubMessage;
+            tryEmitNext(groupMessageEventSink, new GroupMessageEvent(
+                groupDataMessage.getGroup(),
+                groupDataMessage.getData(),
+                groupDataMessage.getDataType(),
+                groupDataMessage.getFromUserId(),
+                groupDataMessage.getSequenceId()));
+
+            if (groupDataMessage.getSequenceId() != null) {
+                sequenceAckId.update(groupDataMessage.getSequenceId());
+            }
+        } else if (webPubSubMessage instanceof ServerDataMessage) {
+            ServerDataMessage serverDataMessage = (ServerDataMessage) webPubSubMessage;
+            tryEmitNext(serverMessageEventSink, new ServerMessageEvent(
+                serverDataMessage.getData(),
+                serverDataMessage.getDataType(),
+                serverDataMessage.getSequenceId()));
+
+            if (serverDataMessage.getSequenceId() != null) {
+                sequenceAckId.update(serverDataMessage.getSequenceId());
+            }
+        } else if (webPubSubMessage instanceof AckMessage) {
+            ackMessageSink.emitNext((AckMessage) webPubSubMessage,
+                emitFailureHandler("Unable to emit GroupMessageEvent"));
```

Comment:
We should only Emit ConnectedEvent the first time a connection receives connected message. Refer 1.2.4: https://github.com/Azure/azure-webpubsub/blob/main/protocols/client/client-spec.md#12-connected

Reply:
Do you mean client may receive ConnectedEvent after resilient JSON recover?

Currently it not show up in my test, but I will see to handle it.

Reply:
Added a `WebPubSubConnection` class, and moved handling of connectionId, sequenceId, and check of only 1 ConnectedEvent and 1 DisconnectedEvent to this class.